File: backend/src/Ahdah.Application/Projects/Contracts/ProjectRequests.cs
using System.ComponentModel.DataAnnotations;
using Ahdah.Application.Access.Contracts;

namespace Ahdah.Application.Projects.Contracts;

public sealed record NewProjectOwnerRequest
{
    [Required]
    [StringLength(200, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public required string OwnerName { get; init; }

    [Required]
    [StringLength(20)]
    [RegularExpression("^\\s*\\+[1-9][0-9]{7,14}\\s*$")]
    public required string PhoneNumber { get; init; }

    [EmailAddress]
    [StringLength(254)]
    public string? Email { get; init; }

    [StringLength(500, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? Address { get; init; }

    [StringLength(1000, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? Notes { get; init; }
}

public sealed record CreateProjectRequest : IValidatableObject
{
    [Required]
    [StringLength(200, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public required string ProjectName { get; init; }

    [Required]
    [StringLength(500, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public required string SiteAddress { get; init; }

    public decimal? Latitude { get; init; }

    public decimal? Longitude { get; init; }

    [StringLength(20)]
    [RegularExpression("^\\s*\\+[1-9][0-9]{7,14}\\s*$")]
    public string? ContactPhoneNumber { get; init; }

    public decimal ContractValue { get; init; }

    public DateOnly ContractDate { get; init; }

    public DateOnly StartDate { get; init; }

    public DateOnly? ExpectedEndDate { get; init; }

    [StringLength(1500, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? Description { get; init; }

    [StringLength(1000, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? Notes { get; init; }

    public Guid? ProjectOwnerId { get; init; }

    public NewProjectOwnerRequest? NewOwner { get; init; }

    public Guid? SupervisorUserId { get; init; }

    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (ContractDate is { } contractDate && contractDate == default)
        {
            yield return new ValidationResult(
                "Contract date is required.",
                [nameof(ContractDate)]);
        }

        if (StartDate is { } startDateValue && startDateValue == default)
        {
            yield return new ValidationResult(
                "Start date is required.",
                [nameof(StartDate)]);
        }

        if (ProjectOwnerId.HasValue == (NewOwner is not null)
            || ProjectOwnerId == Guid.Empty)
        {
            yield return new ValidationResult(
                "Provide exactly one existing projectOwnerId or newOwner.",
                [nameof(ProjectOwnerId), nameof(NewOwner)]);
        }

        if (SupervisorUserId == Guid.Empty)
        {
            yield return new ValidationResult(
                "Supervisor user ID must be a non-empty GUID.",
                [nameof(SupervisorUserId)]);
        }

        if (!ProjectRequestRules.IsContractValueValid(ContractValue))
        {
            yield return new ValidationResult(
                "Contract value must be positive and fit NUMERIC(18,2).",
                [nameof(ContractValue)]);
        }

        if (!ProjectRequestRules.AreCoordinatesValid(Latitude, Longitude))
        {
            yield return new ValidationResult(
                "Latitude and longitude must be supplied together and be in range.",
                [nameof(Latitude), nameof(Longitude)]);
        }

        if (!ProjectRequestRules.AreDatesValid(StartDate, ExpectedEndDate))
        {
            yield return new ValidationResult(
                "Expected end date cannot be before the start date.",
                [nameof(ExpectedEndDate)]);
        }
    }
}

public sealed record UpdateProjectRequest : IValidatableObject
{
    [Range(1, int.MaxValue)]
    public int ExpectedVersion { get; init; }

    [StringLength(200, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? ProjectName { get; init; }

    [StringLength(500, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? SiteAddress { get; init; }

    [StringLength(20)]
    [RegularExpression("^\\s*\\+[1-9][0-9]{7,14}\\s*$")]
    public string? ContactPhoneNumber { get; init; }

    public DateOnly? ContractDate { get; init; }

    public DateOnly? StartDate { get; init; }

    public DateOnly? ExpectedEndDate { get; init; }

    [RegularExpression("^(Active|Paused)$")]
    public string? Status { get; init; }

    [StringLength(1500, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? Description { get; init; }

    [StringLength(1000, MinimumLength = 1)]
    [RegularExpression("^(?=.*\\S)[^\\r\\n]+$")]
    public string? Notes { get; init; }

    public Guid? ProjectOwnerId { get; init; }

    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (ContractDate is { } contractDate && contractDate == default)
        {
            yield return new ValidationResult(
                "Contract date must be valid when supplied.",
                [nameof(ContractDate)]);
        }

        if (StartDate is { } startDateValue && startDateValue == default)
        {
            yield return new ValidationResult(
                "Start date must be valid when supplied.",
                [nameof(StartDate)]);
        }

        if (ProjectOwnerId == Guid.Empty)
        {
            yield return new ValidationResult(
                "Project owner ID must be a non-empty GUID.",
                [nameof(ProjectOwnerId)]);
        }

        if (StartDate is { } startDate
            && ExpectedEndDate is { } expectedEndDate
            && !ProjectRequestRules.AreDatesValid(startDate, expectedEndDate))
        {
            yield return new ValidationResult(
                "Expected end date cannot be before the start date.",
                [nameof(ExpectedEndDate)]);
        }

        if (ProjectName is null
            && SiteAddress is null
            && ContactPhoneNumber is null
            && ContractDate is null
            && StartDate is null
            && ExpectedEndDate is null
            && Status is null
            && Description is null
            && Notes is null
            && ProjectOwnerId is null)
        {
            yield return new ValidationResult("At least one project field must be supplied.");
        }
    }
}

public sealed record AssignProjectSupervisorRequest
{
    public Guid SupervisorUserId { get; init; }

    [Range(1, int.MaxValue)]
    public int ExpectedVersion { get; init; }
}

public sealed record ProjectQuery : PagedAccessQuery
{
    [RegularExpression("^(Active|Paused|Completed|FinanciallyClosed|Cancelled)$")]
    public string? Status { get; init; }

    [StringLength(ProjectConstants.MaximumSearchLength, MinimumLength = 2)]
    [RegularExpression("^[^\\r\\n]+$")]
    public string? Search { get; init; }
}

public sealed record ProjectMemberQuery : PagedAccessQuery;

File: frontend/ahdah_app/lib/features/projects/domain/project_requests.dart
import 'dart:convert';

final class NewProjectOwnerInput {
  const NewProjectOwnerInput({
    required this.ownerName,
    required this.phoneNumber,
    this.email,
    this.address,
    this.notes,
  });

  final String ownerName;
  final String phoneNumber;
  final String? email;
  final String? address;
  final String? notes;

  Map<String, Object?> toJson() => {
    'ownerName': ownerName.trim(),
    'phoneNumber': phoneNumber.trim(),
    'email': _normalizedOptional(email),
    'address': _normalizedOptional(address),
    'notes': _normalizedOptional(notes),
  };
}

final class ProjectCreateInput {
  const ProjectCreateInput({
    required this.projectName,
    required this.siteAddress,
    required this.contractValue,
    required this.contractDate,
    required this.startDate,
    required this.newOwner,
    this.contactPhoneNumber,
    this.expectedEndDate,
    this.description,
    this.notes,
    this.supervisorUserId,
  });

  final String projectName;
  final String siteAddress;
  final String? contactPhoneNumber;
  final String contractValue;
  final DateTime contractDate;
  final DateTime startDate;
  final DateTime? expectedEndDate;
  final String? description;
  final String? notes;
  final NewProjectOwnerInput newOwner;
  final String? supervisorUserId;

  Map<String, Object?> toJson() => {
    'projectName': projectName.trim(),
    'siteAddress': siteAddress.trim(),
    'contactPhoneNumber': _normalizedOptional(contactPhoneNumber),
    'contractValue': contractValue.trim(),
    'contractDate': _date(contractDate),
    'startDate': _date(startDate),
    'expectedEndDate': expectedEndDate == null ? null : _date(expectedEndDate!),
    'description': _normalizedOptional(description),
    'notes': _normalizedOptional(notes),
    'newOwner': newOwner.toJson(),
    'supervisorUserId': _normalizedOptional(supervisorUserId),
  };

  String toJsonBody() {
    final marker = '__AHDAH_VALIDATED_DECIMAL__';
    final body = toJson()..['contractValue'] = marker;
    return jsonEncode(body).replaceFirst('"$marker"', contractValue.trim());
  }
}

final class ProjectUpdateInput {
  const ProjectUpdateInput({
    required this.expectedVersion,
    this.projectName,
    this.siteAddress,
    this.contactPhoneNumber,
    this.contractDate,
    this.startDate,
    this.expectedEndDate,
    this.status,
    this.description,
    this.notes,
  });

  final int expectedVersion;
  final String? projectName;
  final String? siteAddress;
  final String? contactPhoneNumber;
  final DateTime? contractDate;
  final DateTime? startDate;
  final DateTime? expectedEndDate;
  final String? status;
  final String? description;
  final String? notes;

  Map<String, Object?> toJson() => {
    'expectedVersion': expectedVersion,
    if (projectName != null) 'projectName': projectName!.trim(),
    if (siteAddress != null) 'siteAddress': siteAddress!.trim(),
    if (contactPhoneNumber != null)
      'contactPhoneNumber': _normalizedOptional(contactPhoneNumber),
    if (contractDate != null) 'contractDate': _date(contractDate!),
    if (startDate != null) 'startDate': _date(startDate!),
    if (expectedEndDate != null) 'expectedEndDate': _date(expectedEndDate!),
    if (status != null) 'status': status,
    if (description != null) 'description': _normalizedOptional(description),
    if (notes != null) 'notes': _normalizedOptional(notes),
  };
}

final class SupervisorAssignmentInput {
  const SupervisorAssignmentInput({
    required this.supervisorUserId,
    required this.expectedVersion,
  });

  final String supervisorUserId;
  final int expectedVersion;

  Map<String, Object?> toJson() => {
    'supervisorUserId': supervisorUserId,
    'expectedVersion': expectedVersion,
  };
}

String _date(DateTime value) =>
    '${value.year.toString().padLeft(4, '0')}-'
    '${value.month.toString().padLeft(2, '0')}-'
    '${value.day.toString().padLeft(2, '0')}';

String? _normalizedOptional(String? value) {
  final normalized = value?.trim();
  return normalized == null || normalized.isEmpty ? null : normalized;
}
